src/workerd/api/filesystem.c++

```diff
@@ -0,0 +1,1213 @@
+#include "filesystem.h"
+
+#include "blob.h"
+
+#include <workerd/api/streams/standard.h>
+#include <workerd/api/url-standard.h>
+#include <workerd/api/url.h>
+
+#include <kj/encoding.h>
+
+namespace workerd::api {
+
+namespace {
+constexpr kj::StringPtr getNameForFsType(FsType type) {
+  switch (type) {
+    case FsType::FILE:
+      return "file"_kj;
+    case FsType::DIRECTORY:
+      return "directory"_kj;
+    case FsType::SYMLINK:
+      return "symlink"_kj;
+    default:
+      KJ_UNREACHABLE;
+  }
+}
+
+constexpr FsType getTypeForName(kj::StringPtr name) {
+  if (name == "file"_kj) return FsType::FILE;
+  if (name == "directory"_kj) return FsType::DIRECTORY;
+  if (name == "symlink"_kj) return FsType::SYMLINK;
+  KJ_UNREACHABLE;
+}
+
+jsg::Url filePathToUrl(FilePath& path) {
+  KJ_SWITCH_ONEOF(path) {
+    KJ_CASE_ONEOF(str, kj::String) {
+      // For a string, we need to try parsing it as a URL first.
+      auto url = JSG_REQUIRE_NONNULL(jsg::Url::tryParse(str, "file:///"_kj), Error, "Invalid path");
+      JSG_REQUIRE(url.getProtocol() == "file:"_kjb, Error, "Only file: URLs are supported");
```

Comment:
it will always start with file: since the second argument of tryParse is file://. I think this should be a debug assertion rather than a JSG_REQUIRE